tests/dotnet/UnitTests/AssetsTest.cs

```diff
@@ -0,0 +1,218 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Diagnostics;
+using System.Text.Json;
+
+using Mono.Cecil;
+using NUnit.Framework;
+using Xamarin.Utils;
+using Xamarin.Tests;
+using Xamarin.MacDev;
+
+#nullable enable
+
+namespace Xamarin.Tests {
+
+	[TestFixture (ApplePlatform.iOS, "iossimulator-x64", "iphonesimulator")]
+	[TestFixture (ApplePlatform.iOS, "ios-arm64;ios-arm", "iphoneos")]
+	[TestFixture (ApplePlatform.TVOS, "tvossimulator-x64", "appletvsimulator")]
+	[TestFixture (ApplePlatform.MacCatalyst, "maccatalyst-x64", "macosx")]
+	[TestFixture (ApplePlatform.MacCatalyst, "maccatalyst-arm64;maccatalyst-x64", "macosx")]
+	[TestFixture (ApplePlatform.MacOSX, "osx-x64", "macosx")]
+	[TestFixture (ApplePlatform.MacOSX, "osx-arm64;osx-x64", "macosx")] // https://github.com/xamarin/xamarin-macios/issues/12410
+	public class AssetsTest : TestBaseClass {
+
+		readonly ApplePlatform platform;
+		readonly string runtimeIdentifiers;
+		readonly string sdkVersion;
+		string project_path = string.Empty;
+		string appPath = string.Empty;
+
+		public AssetsTest (ApplePlatform platform, string runtimeIdentifiers, string sdkVersion)
+		{
+			this.platform = platform;
+			this.runtimeIdentifiers = runtimeIdentifiers;
+			this.sdkVersion = sdkVersion;
+		}
+
+		[SetUp]
+		public void Init ()
+		{
+			var project = "AppWithXCAssets";
+			Configuration.IgnoreIfIgnoredPlatform (platform);
+			project_path = GetProjectPath (project, runtimeIdentifiers: runtimeIdentifiers, platform: platform, out appPath);
+			DeleteAssets (project_path);
```

Comment:
Why is delete assets in the setup? What is the goal? Should it be in a tear down method too? Unit tests should leave the system as they found it.

Reply:
I thought the point of adding the setup was to delete the assets in the setup and not have to do it inside the tests. Otherwise, I don't see the point in having the setup since it would just be creating the project_path that we need to now use instance variables for. I guess the delete can happen in the tear down as well and then the delete in the setup would just make sure it is not there, the test would add it, and then the clean up would remove it and then it would be the same as before we started the test!

Reply:
The issue is a follows:

1. Call setup. Set the project paths etc, configure the assets (which you are doing in the test)
2. Execute the code to test and assert expectations. Do not do any configuration here, that's part of the setup.
3. Tear down and delete de assets.

I don't understand why you call delete, since nothing has been configured at that point, and any left overs have to be cleaned in the tear down. The 3 steps I have described are the basics of unit testing.

Reply:
An important detail to remember is that the assert methods form nunit use exceptions, that means that the execution of the test stops as soon and there is a failing assertion. 

If the steps that need to clean the code are after the asserts and we have a failure, the clean steps won’t be executed. That forces the setup to be written with the assumption of a dirty workspace which is incorrect.

For reference: https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Assert.cs#L380

That is one of the main reasons I asked you to move to use a setup and a tear down. Not only to make the code much cleaner (as @chamons  wonderfuly explained) but to avoid  tests to have side effects for subsequent tests. What you were experiencing with the previous implementation, in which you did not know if you had to delete a symlink, is rooted in that your tests were not correctly isolated from each other. 

Calling delete assets in the setup was/is a red flag. The same with the method that configured the assets, those should be in the setup method too.

Reply:
@chamons @mandel-macaque AH okay thank you both for the explanations! I will get right on this :)